### PR TITLE
Update ImageSpec documentation

### PR DESCRIPTION
**Pull request by pingsutw**
## What changes were proposed in this pull request?
Update doc and add more examples

Preview: https://flyte--5748.org.readthedocs.build/en/5748/user_guide/customizing_dependencies/imagespec.html

### CODE REVIEW

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 1]
```diff
@@ -6,68 +6,183 @@
 .. tags:: Containerization, Intermediate
 ```
 
-`ImageSpec` is a way to specify how to build a container image without a Dockerfile. The `ImageSpec` by default will be
-converted to an [Envd](https://envd.tensorchord.ai/) config, and the [Envd builder](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-envd/flytekitplugins/envd/image_builder.py#L12-L34) will build the image for you. However, you can also register your own builder to build
-the image using other tools.
+This guide demonstrates how to use the `ImageSpec` to customize the container image for your Flyte tasks.
+
+`ImageSpec` is a mechanism that allows you to specify how to build a container image without requiring a Dockerfile.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:11]:
```suggestion
`ImageSpec` allows you to customize the container image for your Flyte tasks without a Dockerfile.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 2]
```diff
@@ -6,68 +6,183 @@
 .. tags:: Containerization, Intermediate
 ```
 
-`ImageSpec` is a way to specify how to build a container image without a Dockerfile. The `ImageSpec` by default will be
-converted to an [Envd](https://envd.tensorchord.ai/) config, and the [Envd builder](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-envd/flytekitplugins/envd/image_builder.py#L12-L34) will build the image for you. However, you can also register your own builder to build
-the image using other tools.
+This guide demonstrates how to use the `ImageSpec` to customize the container image for your Flyte tasks.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:9]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 3]
```diff
@@ -6,68 +6,183 @@
...
+
+`ImageSpec` is a mechanism that allows you to specify how to build a container image without requiring a Dockerfile.
+This feature provides a flexible and user-friendly way to define the containerization process.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:12]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 4]
```diff
@@ -6,68 +6,183 @@
...
+One of the significant advantages of using ImageSpec is its ability to optimize the image-building process by reusing previously
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:13]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 5]
```diff
@@ -6,68 +6,183 @@
...
+downloaded packages from the PyPI and APT caches, thereby speeding up the entire build process.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:14]:
```suggestion
`ImageSpec` speeds up the build process by allowing you to reuse previously downloaded packages from the PyPI and APT caches.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 6]
```diff
@@ -6,68 +6,183 @@
...
+This efficient caching mechanism minimizes redundant downloads, reducing both the time and resources required to construct the container image.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:15]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 7]
```diff
@@ -6,68 +6,183 @@
...
+
+By default, the `ImageSpec` will be built using the `default` builder associated with Flytekit.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:17]:
```suggestion
By default, the `ImageSpec` will be built using the `default` builder associated with Flytekit, but you can register your own builder.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 8]
```diff
@@ -6,68 +6,183 @@
...
+However, you have the option to register your own builder, enabling you to build the image using alternative tools.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:18]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 9]
```diff
@@ -6,68 +6,183 @@
...
+
+For example, [flytekitplugins-envd](https://github.com/flyteorg/flytekit/blob/c06ef30518dec2057e554fbed375dfa43b985c60/plugins/flytekit-envd/flytekitplugins/envd/image_builder.py#L25) is another image builder that uses envd to build the ImageSpec.
+This flexibility allows you to tailor the image-building process according to your specific needs and tool preferences.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:21]:
```suggestion
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 10]
```diff
@@ -6,68 +6,183 @@
...
-
-```{note}
-To clone and run the example code on this page, see the [Flytesnacks repo][flytesnacks].
-```
-
-```{rli} https://raw.githubusercontent.com/flyteorg/flytesnacks/69dbe4840031a85d79d9ded25f80397c6834752d/examples/customizing_dependencies/customizing_dependencies/image_spec.py
-:caption: customizing_dependencies/image_spec.py
-:lines: 1-4
-```
+Before building the image, Flytekit checks the container registry first to see if the image already exists.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:29]:
```suggestion
Before building the image, Flytekit checks the container registry to see if the image already exists.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 11]
```diff
@@ -6,68 +6,183 @@
...
+By doing so, it avoids having to rebuild the image over and over again. If the image does not exist,
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:30]:
```suggestion
If the image does not exist,
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 12]
```diff
@@ -6,68 +6,183 @@
...
+flytekit will build the image before registering the workflow, and replace the image name in the task template with the newly built image name.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:31]:
```suggestion
Flytekit will build the image before registering the workflow and replace the image name in the task template with the newly built image name.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 13]
```diff
@@ -6,68 +6,183 @@
...
 
-- Install [flytekitplugins-envd](https://github.com/flyteorg/flytekit/tree/master/plugins/flytekit-envd) to build the `ImageSpec`.
-- To build the image on remote machine, check this [doc](https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine).
+- Make sure `docker` is running on your local machine.
 - When using a registry in ImageSpec, `docker login` is required to push the image
 :::
 
-You can specify python packages, apt packages, and environment variables in the `ImageSpec`.
+## Install Python or APT packages
+You can specify python packages and apt packages in the `ImageSpec`.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:41]:
```suggestion
You can specify Python packages and APT packages in the `ImageSpec`.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 14]
```diff
@@ -6,68 +6,183 @@
...
 More specifically, flytekit invokes [DefaultImages.default_image()](https://github.com/flyteorg/flytekit/blob/f2cfef0ec098d4ae8f042ab915b0b30d524092c6/flytekit/configuration/default_images.py#L26-L27) function.
-This function determines and returns the default image based on the Python version and flytekit version. For example, if you are using python 3.8 and flytekit 0.16.0, the default image assigned will be `ghcr.io/flyteorg/flytekit:py3.8-1.6.0`.
-If desired, you can also override the default image by providing a custom `base_image` parameter when using the `ImageSpec`.
-
-```{rli} https://raw.githubusercontent.com/flyteorg/flytesnacks/69dbe4840031a85d79d9ded25f80397c6834752d/examples/customizing_dependencies/customizing_dependencies/image_spec.py
-:caption: customizing_dependencies/image_spec.py
-:lines: 6-19
-```
+This function determines and returns the default image based on the Python version and flytekit version.
+For example, if you are using python 3.8 and flytekit 1.6.0, the default image assigned will be `ghcr.io/flyteorg/flytekit:py3.8-1.6.0`.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:45]:
```suggestion
For example, if you are using Python 3.8 and flytekit 1.6.0, the default image assigned will be `ghcr.io/flyteorg/flytekit:py3.8-1.6.0`.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 15]
```diff
@@ -6,68 +6,183 @@
...
+  name="my-image",
+  packages=["pandas"],
+  tag_format="{spec_hash}-dev",
+  registry="ghcr.io/flyteorg",
+)
 ```
 
-There exists an option to override the container image by providing an Image Spec YAML file to the `pyflyte run` or `pyflyte register` command.
+## Define ImageSpec in a YAML File
+There is an option to override the container image by providing an ImageSpec YAML file to the  `pyflyte run` or `pyflyte register` command.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:185]:
```suggestion

You can override the container image by providing an ImageSpec YAML file to the  `pyflyte run` or `pyflyte register` command.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 16]
```diff
@@ -85,19 +200,22 @@ env:
 pyflyte run --remote --image image.yaml image_spec.py wf
 ```
 
+## Build the image without registering the workflow
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:203]:
```suggestion
## Build the image without registering the workflow

```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 17]
```diff
@@ -85,19 +200,22 @@ env:
...
 If you only want to build the image without registering the workflow, you can use the `pyflyte build` command.
 
 ```
 pyflyte build --remote image_spec.py wf
 ```
 
-In some cases, you may want to force an image to rebuild, even if the image spec hasn’t changed. If you want to overwrite an existing image, you can pass the `FLYTE_FORCE_PUSH_IMAGE_SPEC=True` to `pyflyte` command or add `force_push()` to the ImageSpec.
+## Force Push an Image
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:210]:
```suggestion
## Force push an image

```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 18]
```diff
@@ -85,19 +200,22 @@ env:
...
+In some cases, you may want to force an image to rebuild, even if the image spec hasn’t changed.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:211]:
```suggestion
In some cases, you may want to force an image to rebuild, even if the ImageSpec hasn’t changed.
```

File: docs/user_guide/customizing_dependencies/imagespec.md [hunk 19]
```diff
@@ -85,19 +200,22 @@ env:
...
+If you want to overwrite an existing image, you can pass the `FLYTE_FORCE_PUSH_IMAGE_SPEC=True` to `pyflyte` command.
```
Comment (neverett) [inline docs/user_guide/customizing_dependencies/imagespec.md:212]:
```suggestion
To overwrite an existing image, pass the `FLYTE_FORCE_PUSH_IMAGE_SPEC=True` to the `pyflyte` command.
```

**Review by neverett — CHANGES_REQUESTED** [review]
Some changes requested, otherwise LGTM!

**Comment by davidmirror-ops** [comment]
Thank you @pingsutw! This is going to be very helpful

**Comment by pingsutw** [comment]
@neverett thanks for the review. I've updated the PR. Mind taking a look again.

**Review by neverett — APPROVED** [review]

**Pull request merged**